File: backend/api/simulation.py
"""Simulation API endpoints."""
from __future__ import annotations

import json
from pathlib import Path

from fastapi import APIRouter
from fastapi.responses import StreamingResponse
from pydantic import BaseModel as PydanticModel

from code.config import root
from code.backend.models.schemas import SimulationSession
from code.backend.services.simulation import config_service, result_service
from code.backend.services.simulation.simulation_engine import run_simulation

router = APIRouter(prefix="/api/simulation", tags=["simulation"])

_ROOT = Path(root).resolve()
_DATA_DIR = Path(__file__).resolve().parent.parent / "data"


def _sse(event_type: str, data: dict | None = None) -> str:
    payload = {"type": event_type}
    if data:
        payload.update(data)
    return f"data: {json.dumps(payload, ensure_ascii=False)}\n\n"


# ── Config endpoints ──────────────────────────────────────────────────────────

@router.get("/configs")
def list_configs():
    return {"configs": config_service.list_configs(_ROOT)}


@router.post("/configs/default/update")
def update_default_config():
    """SSE stream: generates thought flows for all patterns in analysis/pattern/."""
    import queue
    import threading

    q: queue.Queue = queue.Queue()

    def emit(s: str) -> None:
        q.put(s)

    def worker():
        try:
            config_service.update_default_config(_ROOT, emit=emit)
        except Exception as e:
            q.put(_sse("error", {"message": str(e)}))
        finally:
            q.put(None)  # sentinel

    threading.Thread(target=worker, daemon=True).start()

    def gen():
        while True:
            item = q.get()
            if item is None:
                yield _sse("done", {"message": "默认配置更新完成"})
                break
            yield item

    return StreamingResponse(gen(), media_type="text/event-stream", headers={
        "Cache-Control": "no-cache",
        "Connection": "keep-alive",
        "X-Accel-Buffering": "no",
    })


class RenameConfigPayload(PydanticModel):
    new_name: str


@router.post("/configs")
def create_config():
    new_name = config_service.create_config(_ROOT)
    return {"config_name": new_name}


@router.put("/configs/{config_name}/rename")
def rename_config(config_name: str, payload: RenameConfigPayload):
    try:
        config_service.rename_config(_ROOT, config_name, payload.new_name)
    except (FileNotFoundError, ValueError) as e:
        from fastapi import HTTPException
        raise HTTPException(status_code=400, detail=str(e))
    return {"config_name": payload.new_name}


@router.get("/configs/{config_name}/patterns")
def list_patterns(config_name: str):
    patterns = config_service.list_patterns(_ROOT, config_name)
    return {"patterns": [p.model_dump() for p in patterns]}


class UpdateIntensityPayload(PydanticModel):
    rel_path: str
    intensity: float


@router.put("/configs/{config_name}/patterns/intensity")
def update_intensity(config_name: str, payload: UpdateIntensityPayload):
    config_service.update_pattern_intensity(_ROOT, config_name, payload.rel_path, payload.intensity)
    return {"success": True}


# ── Simulation run ────────────────────────────────────────────────────────────

class RunSimulationPayload(PydanticModel):
    session_id: str
    config_name: str
    situation: str
    max_branches: int = 2
    max_steps: int = 3
    alpha: float = 0.6
    seed: int | None = None


@router.post("/run")
def run_simulation_endpoint(payload: RunSimulationPayload):
    import queue
    import threading

    q: queue.Queue = queue.Queue()

    def emit(s: str) -> None:
        q.put(s)

    def worker():
        try:
            patterns = config_service.list_patterns(_ROOT, payload.config_name)
            result = run_simulation(
                situation=payload.situation,
                config_name=payload.config_name,
                patterns=patterns,
                root=_ROOT,
                max_branches=payload.max_branches,
                max_steps=payload.max_steps,
                alpha=payload.alpha,
                emit=emit,
                seed=payload.seed,
            )
            result_service.save_result(_ROOT, result)
            result_service.attach_result_to_session(_DATA_DIR, payload.session_id, result.id)
        except Exception as e:
            q.put(_sse("error", {"message": str(e)}))
        finally:
            q.put(None)

    threading.Thread(target=worker, daemon=True).start()

    def gen():
        while True:
            item = q.get()
            if item is None:
                break
            yield item

    return StreamingResponse(gen(), media_type="text/event-stream", headers={
        "Cache-Control": "no-cache",
        "Connection": "keep-alive",
        "X-Accel-Buffering": "no",
    })


# ── Session endpoints ─────────────────────────────────────────────────────────

@router.get("/sessions")
def list_sessions():
    sessions = result_service.list_sessions(_DATA_DIR)
    return {"sessions": [s.model_dump() for s in sessions]}


class CreateSessionPayload(PydanticModel):
    name: str


@router.post("/sessions")
def create_session(payload: CreateSessionPayload):
    session = result_service.create_session(_DATA_DIR, payload.name)
    return session.model_dump()


class RenameSessionPayload(PydanticModel):
    name: str


@router.put("/sessions/{session_id}/rename")
def rename_session(session_id: str, payload: RenameSessionPayload):
    session = result_service.rename_session(_DATA_DIR, session_id, payload.name)
    return session.model_dump()


@router.delete("/sessions/{session_id}")
def delete_session(session_id: str):
    result_service.delete_session(_DATA_DIR, session_id, root=_ROOT)
    return {"success": True}


# ── Result endpoints ──────────────────────────────────────────────────────────

@router.get("/results/{result_id}")
def get_result(result_id: str):
    result = result_service.get_result(_ROOT, result_id)
    return result.model_dump()

File: code/frontend/src/api/simulation.ts
import type { PatternItem, SimulationResult, SimulationSession, SimSSEEvent } from "../types/simulation"

const BASE = "/api/simulation"

export async function listConfigs(): Promise<string[]> {
  const res = await fetch(`${BASE}/configs`)
  if (!res.ok) throw new Error(await res.text())
  const data = await res.json()
  return data.configs
}

export async function updateDefaultConfig(
  onEvent: (e: SimSSEEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  const res = await fetch(`${BASE}/configs/default/update`, { method: "POST", signal })
  if (!res.ok) throw new Error(await res.text())
  await consumeSSE(res, onEvent)
}

export async function renameConfig(oldName: string, newName: string): Promise<string> {
  const res = await fetch(`${BASE}/configs/${encodeURIComponent(oldName)}/rename`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ new_name: newName }),
  })
  if (!res.ok) throw new Error(await res.text())
  const data = await res.json()
  return data.config_name
}

export async function createConfig(): Promise<string> {
  const res = await fetch(`${BASE}/configs`, { method: "POST" })
  if (!res.ok) throw new Error(await res.text())
  const data = await res.json()
  return data.config_name
}

export async function listPatterns(configName: string): Promise<PatternItem[]> {
  const res = await fetch(`${BASE}/configs/${encodeURIComponent(configName)}/patterns`)
  if (!res.ok) throw new Error(await res.text())
  const data = await res.json()
  return data.patterns
}

export async function updatePatternIntensity(
  configName: string,
  relPath: string,
  intensity: number,
): Promise<void> {
  const res = await fetch(`${BASE}/configs/${encodeURIComponent(configName)}/patterns/intensity`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ rel_path: relPath, intensity }),
  })
  if (!res.ok) throw new Error(await res.text())
}

export async function runSimulation(
  params: {
    session_id: string
    config_name: string
    situation: string
    max_branches: number
    max_steps: number
    alpha: number
  },
  onEvent: (e: SimSSEEvent) => void,
  signal?: AbortSignal,
): Promise<void> {
  const res = await fetch(`${BASE}/run`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(params),
    signal,
  })
  if (!res.ok) throw new Error(await res.text())
  await consumeSSE(res, onEvent)
}

export async function listSessions(): Promise<SimulationSession[]> {
  const res = await fetch(`${BASE}/sessions`)
  if (!res.ok) throw new Error(await res.text())
  const data = await res.json()
  return data.sessions
}

export async function createSession(name: string): Promise<SimulationSession> {
  const res = await fetch(`${BASE}/sessions`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) throw new Error(await res.text())
  return res.json()
}

export async function renameSession(sessionId: string, name: string): Promise<SimulationSession> {
  const res = await fetch(`${BASE}/sessions/${sessionId}/rename`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ name }),
  })
  if (!res.ok) throw new Error(await res.text())
  return res.json()
}

export async function deleteSession(sessionId: string): Promise<void> {
  const res = await fetch(`${BASE}/sessions/${sessionId}`, { method: "DELETE" })
  if (!res.ok) throw new Error(await res.text())
}

export async function getResult(resultId: string): Promise<SimulationResult> {
  const res = await fetch(`${BASE}/results/${resultId}`)
  if (!res.ok) throw new Error(await res.text())
  return res.json()
}

async function consumeSSE(res: Response, onEvent: (e: SimSSEEvent) => void): Promise<void> {
  const reader = res.body?.getReader()
  if (!reader) return
  const decoder = new TextDecoder()
  let buf = ""
  while (true) {
    const { done, value } = await reader.read()
    if (done) break
    buf += decoder.decode(value, { stream: true })
    const lines = buf.split("\n")
    buf = lines.pop() ?? ""
    for (const line of lines) {
      if (line.startsWith("data: ")) {
        try {
          const event = JSON.parse(line.slice(6)) as SimSSEEvent
          onEvent(event)
        } catch { /* ignore */ }
      }
    }
  }
}
